vault/acl.go

```diff
@@ -47,25 +47,83 @@ func NewACL(policies []*Policy) (*ACL, error) {
 			// Check for an existing policy
 			raw, ok := tree.Get(pc.Prefix)
 			if !ok {
-				tree.Insert(pc.Prefix, pc.CapabilitiesBitmap)
+				tree.Insert(pc.Prefix, pc.Permissions)
 				continue
 			}
-			existing := raw.(uint32)
+
+			// these are the ones already in the tree
+			permissions := raw.(*Permissions)
+			existing := permissions.CapabilitiesBitmap
 
 			switch {
 			case existing&DenyCapabilityInt > 0:
 				// If we are explicitly denied in the existing capability set,
 				// don't save anything else
 
-			case pc.CapabilitiesBitmap&DenyCapabilityInt > 0:
+			case pc.Permissions.CapabilitiesBitmap&DenyCapabilityInt > 0:
 				// If this new policy explicitly denies, only save the deny value
-				tree.Insert(pc.Prefix, DenyCapabilityInt)
+				pc.Permissions.CapabilitiesBitmap = DenyCapabilityInt
+				tree.Insert(pc.Prefix, pc.Permissions)
 
 			default:
 				// Insert the capabilities in this new policy into the existing
 				// value
-				tree.Insert(pc.Prefix, existing|pc.CapabilitiesBitmap)
+				pc.Permissions.CapabilitiesBitmap = existing | pc.Permissions.CapabilitiesBitmap
+				tree.Insert(pc.Prefix, pc.Permissions)
+			}
+
+			// look for a * in allowed parameters for the node already in the tree
+			if _, ok := permissions.AllowedParameters["*"]; ok {
+				pc.Permissions.AllowedParameters = make(map[string]struct{})
+				pc.Permissions.AllowedParameters["*"] = struct{}{}
+				goto CHECK_DENIED
+			}
+
+			// look for a * in allowed parameters for the path capability we are merging
+			if _, ok := pc.Permissions.AllowedParameters["*"]; ok {
+				pc.Permissions.AllowedParameters = make(map[string]struct{})
```

Comment:
In this case you don't need to make a new map, because if you modify the check below as I suggest, you'll have only * in it anyways, so you can just goto. However, you should be checking that AllowedParameters is not nil before this. See my comment below.